packages/accounts-2fa/2fa-server.js

```diff
@@ -0,0 +1,132 @@
+import { Accounts } from 'meteor/accounts-base';
+import twofactor from 'node-2fa';
+import QRCode from 'qrcode-svg';
+import { Meteor } from 'meteor/meteor';
+
+Accounts._is2faEnabledForUser = selector => {
+  if (!Meteor.isServer) {
+    throw new Meteor.Error(
+      400,
+      'The function _is2faEnabledForUser can only be called on the server'
+    );
+  }
+
+  if (typeof selector === 'string') {
+    if (!selector.includes('@')) {
+      selector = { username: selector };
+    } else {
+      selector = { email: selector };
+    }
+  }
+
+  const user = Meteor.users.findOne(selector) || {};
+  const { services: { twoFactorAuthentication } = {} } = user;
+  return (
+    twoFactorAuthentication &&
+    twoFactorAuthentication.secret &&
+    twoFactorAuthentication.type === 'otp'
+  );
+};
+
+Accounts._generate2faToken = secret => twofactor.generateToken(secret);
+
+Accounts._isTokenValid = (secret, code) => {
+  if (!Meteor.isServer) {
+    throw new Meteor.Error(
+      400,
+      'The function _isTokenValid can only be called on the server'
+    );
+  }
+  const { delta } = twofactor.verifyToken(secret, code, 10) || {};
+  return delta != null && delta >= 0;
+};
+
+Meteor.methods({
+  generate2faActivationQrCode(appName) {
+    const user = Meteor.user();
+
+    if (!user) {
+      throw new Meteor.Error(
+        400,
+        'There must be a user logged in to generate the QR code.'
+      );
+    }
+
+    const { username } = user;
+
+    const { secret, uri } = twofactor.generateSecret({
+      name: appName.trim(),
+      account: username,
+    });
+    const svg = new QRCode(uri).svg();
+
+    Meteor.users.update(
+      { username },
```

Comment:
All the selectors should primarily be by `_id`, username is an optional field and I have seen use cases where username is not assigned to user.

Reply:
I'll try to get a proposed fix by the end of today.